Introduce fast path for innerHTML

cdumez opened the pull request:
#### 236ad6cad0564c857858202c291d383ec4a31a60
<pre>
Introduce fast path for innerHTML
<a href="https://bugs.webkit.org/show_bug.cgi?id=252054">https://bugs.webkit.org/show_bug.cgi?id=252054</a>

Reviewed by Darin Adler.

This patch imports the new fast HTML parser from Blink:
<a href="https://bugs.chromium.org/p/chromium/issues/detail?id=1407201">https://bugs.chromium.org/p/chromium/issues/detail?id=1407201</a>

This patch adds a new parser that handles a subset of HTML and is only used by
Element::set[Inner|Outer]HTML. While limited, the hope is this parser is
useful for a number of pages. We could expand on the set of cases this handles
in the future.

As a sanity check, there is a debug assertion in place to make sure that the
fast path generates the same output as the full HTML parser would have.

This is a verified 2.25% progression on Speedometer on macOS and 2.60%
progression on iOS.

* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.h:
(WebCore::Document::isDirAttributeDirty const):
(WebCore::Document::setIsDirAttributeDirty):
(WebCore::Document::isTemplateDocument const):
* Source/WebCore/editing/markup.cpp:
(WebCore::createFragmentForMarkup):
* Source/WebCore/html/HTMLBDIElement.cpp:
(WebCore::HTMLBDIElement::HTMLBDIElement):
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::create):
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::childrenChanged):
(WebCore::HTMLElement::dirAttributeChanged):

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::~HTMLInputElement):
(WebCore::HTMLInputElement::needsSuspensionCallback):
We have an optimization in HTMLInputElement which delays the initialization of
m_inputType when constructed by the HTML parser. However, in the fast-parsing
case, we may now fail parsing (due to limited support) in the middle of parsing
an HTMLInputElement. As a result, we may end up destroying an HTMLInputElement
before its m_inputType gets initialized, which would cause a crash. I added a
couple of null checks for m_inputType so that the HTMLInputElement destructor
can run even if m_inputType wasn&apos;t initialized yet.

* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::create):
* Source/WebCore/html/HTMLLabelElement.h:
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::create):
* Source/WebCore/html/HTMLSelectElement.h:
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp: Added.
(WebCore::operator==):
(WebCore::onlyContainsLowercaseASCIILetters):
(WebCore::tagNameHash):
(WebCore::HTMLFastPathParser::HTMLFastPathParser):
(WebCore::HTMLFastPathParser::parse):
(WebCore::HTMLFastPathParser::parseResult const):
(WebCore::HTMLFastPathParser::TagInfo::Tag::create):
(WebCore::HTMLFastPathParser::TagInfo::Tag::allowedInPhrasingOrFlowContent):
(WebCore::HTMLFastPathParser::TagInfo::Tag::allowedInFlowContent):
(WebCore::HTMLFastPathParser::TagInfo::ContainerTag::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::ContainsPhrasingContentTag::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::A::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::AWithPhrasingContent::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::B::create):
(WebCore::HTMLFastPathParser::TagInfo::Footer::create):
(WebCore::HTMLFastPathParser::TagInfo::I::create):
(WebCore::HTMLFastPathParser::TagInfo::Input::create):
(WebCore::HTMLFastPathParser::TagInfo::Option::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::Ol::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::Select::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::Strong::create):
(WebCore::HTMLFastPathParser::TagInfo::Ul::parseChild):
(WebCore::HTMLFastPathParser::parseCompleteInput):
(WebCore::HTMLFastPathParser::isWhitespace):
(WebCore::HTMLFastPathParser::isValidUnquotedAttributeValueChar):
(WebCore::HTMLFastPathParser::isValidAttributeNameChar):
(WebCore::HTMLFastPathParser::isCharAfterTagNameOrAttribute):
(WebCore::HTMLFastPathParser::isCharAfterUnquotedAttribute):
(WebCore::HTMLFastPathParser::skipWhitespace):
(WebCore::HTMLFastPathParser::scanText):
(WebCore::HTMLFastPathParser::scanEscapedText):
(WebCore::HTMLFastPathParser::scanTagName):
(WebCore::HTMLFastPathParser::scanAttributeName):
(WebCore::HTMLFastPathParser::scanAttributeValue):
(WebCore::HTMLFastPathParser::scanEscapedAttributeValue):
(WebCore::HTMLFastPathParser::scanHTMLCharacterReference):
(WebCore::HTMLFastPathParser::didFail):
(WebCore::HTMLFastPathParser::peekNext):
(WebCore::HTMLFastPathParser::consumeNext):
(WebCore::HTMLFastPathParser::parseChildren):
(WebCore::HTMLFastPathParser::processAttribute):
(WebCore::HTMLFastPathParser::parseAttributes):
(WebCore::HTMLFastPathParser::parseSpecificElements):
(WebCore::HTMLFastPathParser::parseElement):
(WebCore::HTMLFastPathParser::parseElementAfterTagName):
(WebCore::HTMLFastPathParser::parseContainerElement):
(WebCore::HTMLFastPathParser::parseVoidElement):
(WebCore::canUseFastPath):
(WebCore::tryFastParsingHTMLFragmentImpl):
(WebCore::tryFastParsingHTMLFragment):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.h: Added.
* Source/WebCore/html/parser/HTMLEntityParser.cpp:
(WebCore::appendLegalEntityFor):
* Source/WebCore/html/parser/HTMLEntityParser.h:
* Source/WebCore/html/parser/HTMLNameCache.h:
(WebCore::HTMLNameCache::makeAttributeQualifiedName):
(WebCore::HTMLNameCache::makeQualifiedName):

Canonical link: <a href="https://commits.webkit.org/260148@main">https://commits.webkit.org/260148@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba5d5d40b30b3ed5a8e882a38ab746e86b57cfea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107286 "8 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16343 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40073 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116443 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111185 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17769 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7581 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99439 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113051 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96486 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/41023 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95311 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/28122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96664 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9378 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29464 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/96095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7369 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10019 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/96095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15533 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49043 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104924 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7020 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11571 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25998 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -104,6 +104,8 @@ class String final {
     unsigned length() const { return m_impl ? m_impl->length() : 0; }
     const LChar* characters8() const { return m_impl ? m_impl->characters8() : nullptr; }
     const UChar* characters16() const { return m_impl ? m_impl->characters16() : nullptr; }
+    Span<const LChar> span8() const { return { characters8(), length() }; }
```
Kinda feel like these should have the characters8/16 name long term and the raw pointers one can either go away or have a longer name. Not worth changing in this patch, but maybe worth a comment.

cdumez replied:
I tried to be consistent with StringView which already had those:
```
    const LChar* characters8() const;
    const UChar* characters16() const;
    Span<const LChar> span8() const { return { characters8(), length() }; }
    Span<const UChar> span16() const { return { characters16(), length() }; }
```

darinadler replied:
For future consideration, I like Sam’s idea even though I am the one who named the functions in `StringView`.

weinig commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+    Document& m_document;
+    DocumentFragment& m_fragment;
+
+    const Char* const m_end { m_source.data() + m_source.size() };
```
Did you consider using an abstraction like the StringParsingBuffer rather than raw pointers?

cdumez replied:
This is a straight import. I didn't make any tweaks yet. I think we should adopt more WebKit concepts, though I was hoping to do it in a follow-up since each change will have to be perf-tested.

cdumez replied:
I'll give it a shot in a follow-up and perf test it. I would be nicer than dealing with pointers indeed.

cdumez replied:
I was able to rewrite the code using StringParsingBuffer easily. I am waiting on A/B perf results.

cdumez replied:
Looks like a small regression at the moment, I am trying to figure out why.

darinadler commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -1345,7 +1346,16 @@ static ALWAYS_INLINE ExceptionOr<Ref<DocumentFragment>> createFragmentForMarkup(
     ASSERT(!fragment->hasChildNodes());
     if (document->isHTMLDocument()) {
-        fragment->parseHTML(markup, &contextElement, parserContentPolicy);
+        bool parsedUsingFastPath = tryFastParsingHTMLFragment(markup, document, fragment, contextElement, parserContentPolicy);
+        if (parsedUsingFastPath) {
+#if ASSERT_ENABLED
+            // As a sanity check for the fast-path, create another fragment using the full parser and compare the results.
+            auto referenceFragment = DocumentFragment::create(document);
+            referenceFragment->parseHTML(markup, &contextElement, parserContentPolicy);
+            ASSERT(serializeFragment(fragment, SerializedNodes::SubtreesOfChildren) == serializeFragment(referenceFragment, SerializedNodes::SubtreesOfChildren));
+#endif
+        } else
+            fragment->parseHTML(markup, &contextElement, parserContentPolicy);
```
Could this all go into the `fragment->parseHTML` function, not here? The abstraction is "faster way to parse", not "two different ways to parse", so I think one level deeper.

cdumez replied:
The problem is that the fast path can start parsing and populating the fragment, then encounter something it can't parse and give up. When it gives up we currently destroy the fragment and reconstruct a new one before calling `fragment->parseHTML()`. This reconstruction makes what you suggest not possible so we'd need another approach. I already tried calling `removeChildren()` on the fragment but that turned out to be too expensive. I need to find a cheaper way.

darinadler commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -1487,6 +1487,9 @@ bool HTMLInputElement::isOutOfRange() const
 
 bool HTMLInputElement::needsSuspensionCallback()
 {
+    if (!m_inputType)
+        return false;
```
What is this change for? Change log doesn’t say.

cdumez replied:
Same as above.

darinadler commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -141,7 +141,7 @@ HTMLInputElement::~HTMLInputElement()
     // actually adds the button to the document groups in the latter case.
     // That is inelegant, but harmless since we remove it here.
-    if (isRadioButton())
+    if (m_inputType && isRadioButton())
```
What is this for? Change log doesn’t say.

cdumez replied:
I'll add in the commit log. A while back, we landed a Speedometer optimization in HTMLInputElement which delays initialization of the inputType when constructed by the parser. This was to avoid creating an inputType for "text" only to figure out later during parsing that we wanted another type.

There is a case in the new HTML fast parsing case where we give up parsing a destroy a newly constructed HTMLInputElement before we're done fully parsing it. As a result, the input element may now get destroyed *before* its m_inputType got initialized. However, the HTMLInputElement would previously crash if m_inputType is null. It seemed like a waste to force the initialization of m_inputType in this case so I decided to go the other way and add support for an HTMLInputElement to get destroyed before its m_inputType gets initialized.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+    FailedMaxDepth,
+    FailedBigText,
+    FailedCssPseudoDirEnabledAndDirAttributeDirty,
+    MaxValue = FailedCssPseudoDirEnabledAndDirAttributeDirty,
```
We should leave this `MaxValue` thing out since it’s never used.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+namespace WebCore {
+
+// Captures the potential outcomes for fast path html parser.
+// These values are persisted to logs. Entries should not be renumbered and
```
This comment doesn’t seem to be accurate and not sure I agree with this approach.

cdumez replied:
Right, I should update this comment. Blink uses this enum this way. We don't actually currently use this enum but we do have it available when fast-parsing finishes (the parser can tell your why parsing failed using this enum). I figured I would keep this logic & enum so we can more easily add logging to figure out why fast-parsing failed.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+{
+    // The fast-path parser only scans for letters in tagNames.
+    ASSERT(onlyContainsLowercaseASCIILetters<n>(s));
+    ASSERT(s[n - 1] == '\0');
```
Surprised that these don’t need to use `ASSERT_IN_CONSTEXPR_CONTEXT`.

cdumez replied:
It does build without it and I admit I hadn't heard of `ASSERT_UNDER_CONSTEXPR_CONTEXT()` before. I'll try switching to `ASSERT_UNDER_CONSTEXPR_CONTEXT()`

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+                    return !m_parsingFailed;                                                                 \
+                }                                                                                            \
+            }                                                                                                \
+        break;
```
This is indented incorrectly.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+    }
+
+    // Match ASCII Whitespace according to https://infra.spec.whatwg.org/#ascii-whitespace
+    bool isWhitespace(Char c)
```
This looks like a slower version of `isHTMLSpace`, so we could use the real one.

cdumez replied:
OK, I hope our version really is faster.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+
+    bool isValidUnquotedAttributeValueChar(Char c)
+    {
+        return ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z') || ('0' <= c && c <= '9') || c == '_' || c == '-';
```
Seems to include a less optimized `isASCIIAlphanumeric` in here.

cdumez replied:
It seems to allow `_` and '-' which `isASCIIAlphanumeric` wouldn't. We may be able to use `isASCIIAlphanumeric(c) || c == '_' || c == '-'`. Do you mind if I try this in a follow-up. I really should perf test before making such changes.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+    {
+        if (c == '=') // Early return for the most common way to end an attribute.
+            return false;
+        return ('a' <= c && c <= 'z') || c == '-' || ('A' <= c && c <= 'Z') || ('0' <= c && c <= '9');
```
Here too.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+
+    Attribute processAttribute(CharSpan nameSpan, std::pair<CharSpan, UCharSpan> valueSpan)
+    {
+        QualifiedName name = HTMLNameCache::makeAttributeQualifiedName(nameSpan);
```
Why can’t this just call `makeQualifiedName`? Is there some kind of abstraction involved?

cdumez replied:
I couldn't find `makeQualifiedName()`? Unless you mean the private function on `HTMLNameCache`? It is private so I can't currently call it. The HTMLNameCache was introduced fairly recently by Cameron I think as a Speedometer optimization. I matched what the regular parser is doing.

darinadler commented on Source/WebCore/html/parser/HTMLNameCache.h:
```diff
@@ -73,7 +78,8 @@ class HTMLNameCache {
 
-    ALWAYS_INLINE static QualifiedName makeQualifiedName(Span<const UChar> string)
+    template<typename CharType>
+    ALWAYS_INLINE static QualifiedName makeQualifiedName(Span<const CharType> string)
```
Please use `CharacterType` instead of `CharType`.

darinadler commented on Source/WebCore/html/parser/HTMLEntityParser.cpp:
```diff
@@ -146,4 +146,9 @@ size_t decodeNamedEntityToUCharArray(const char* name, UChar result[4])
     return numberOfCodePoints + appendUChar32ToUCharArray(search.mostRecentMatch()->secondValue, result + numberOfCodePoints);
 }
 
+void appendLegalEntityFor(UChar32 c, StringBuilder& decodedEntity)
```
Please use the name `character` rather than `c`.

cdumez replied:
OK.

Is it because it is a `UChar32`?

You didn't make the same comment for the other functions taking a `c` so I started wondering.

darinadler replied:
I just ran out of steam reviewing, really. I think occasionally a single letter name could be justified but in a short function I think we’d want to used a word instead of a letter.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+            StringBuilder entity;
+            appendLegalEntityFor(res, entity);
+            for (unsigned i = 0; i < entity.length(); ++i)
+                out.append(entity[i]);
```
This seems inefficient. Why not just have appendLegalEntityFor take a `Vector<UChar>`? Making and destroying a string each time won’t be fast.

cdumez replied:
Oh good point! I'll make this change. I only used StringBuilder for consistency with the other API in that header but didn't think enough about the performance implications.

cdumez replied:
Planning on going with:
```
void appendLegalEntityFor(UChar32 c, Vector<UChar>& decodedEntity)
{
    auto legalEntity = HTMLEntityParser::legalEntityFor(c);
    if (U_IS_BMP(legalEntity)) {
        decodedEntity.append(static_cast<UChar>(legalEntity));
        return;
    }
    decodedEntity.append(U16_LEAD(legalEntity));
    decodedEntity.append(U16_TRAIL(legalEntity));
}
```

darinadler commented on Source/WebCore/html/parser/HTMLEntityParser.h:
```diff
@@ -31,6 +31,7 @@
 namespace WebCore {
 
 bool consumeHTMLEntity(SegmentedString&, StringBuilder& decodedEntity, bool& notEnoughCharacters, UChar additionalAllowedCharacter = '\0');
+void appendLegalEntityFor(UChar32, StringBuilder& decodedEntity);
```
The name `decodedEntity` is not a good name for a `StringBuilder` that we will be appending to. Sure, the thing we are appending is a decoded entity.

cdumez replied:
Ok. Proposing `ouputBuffer`

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+        m_parsingFailed = true;
+    }
+
+    template<class ResultType> ResultType didFail(HTMLFastPathResult result, ResultType res)
```
These argument names: `result` and `res`, not very helpful. I suggest `returnValue` instead of `res`.

cdumez replied:
Will use `ResultType` -> `ReturnValueType` for the template parameter as well, to avoid confusion.

darinadler commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+                return didFail(HTMLFastPathResult::FailedParsingAttributes);
+            m_attributeNames.append(attribute.localName().impl());
+        }
+        std::sort(m_attributeNames.begin(), m_attributeNames.end());
+        if (std::adjacent_find(m_attributeNames.begin(), m_attributeNames.end()) != m_attributeNames.end()) {
```
Is sorting really the fastest way to find duplicates? Seems unlikely. I think a `HashSet<AtomString>` would be faster.

cdumez replied:
Ok. I'll add a FIXME comment for now and evaluate this in a follow-up.

cdumez replied:
Well, interestingly, switching to a HashSet appears to be a confirmed 0.8% regression on Speedometer on iOS.
My bet at the moment is that the issue with the HashSet is that it we don't have a way to clear it while retaining its internal capacity. Unlike a Vector where you can resize(0) and drop its items while retaining its capacity. This is helpful since the parser currently reuses the same parser between elements.

darinadler replied:
I think we have varieties of `HashSet` including ones with minimum capacity so we don't spend so much time reallocating memory. If the regression really about the memory allocation and not the hashing I think we can fix it. But maybe this is already fast enough.

darinadler replied:
But I guess all of them lack a `clear` function that clears in place. I do think a version that only grows and never shrinks could be useful in many places.

cdumez replied:
I am A/B testing a version with a HashSet that doesn't clear capacity so see if it is faster. If not, I guess the cost would be the hashing.

cdumez replied:
So the version that uses a HashSet and clears it while maintaining its capacity is perf-neutral. It's a little disappointing.
We may still want to do it since a HashSet would be less prone to pathological cases than vector sorting & deduping.

darinadler reviewed (approved):
So many coding style tweaks possibilities, and maybe additional optimization too. But it’s great to get this in here.

cdumez commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -0,0 +1,1053 @@
+    // Used if the attribute name contains upper case ascii (which must be mapped to lower case).
+    // 32 matches that used by HTMLToken::Attribute.
+    Vector<Char, 32> m_attributeNameBuffer;
+    Vector<Attribute> m_attributeBuffer;
```
FYI, this vector is a big contributor to fastMalloc() in the profiles. I think ideally we'd use `Vector<Attribute, 10>` like we do in the full HTML parser. However, this requires some code changes to make it build/work. I am currently A/B testing such change.

cdumez committed:
Introduce fast path for innerHTML
https://bugs.webkit.org/show_bug.cgi?id=252054

Reviewed by Darin Adler.

This patch imports the new fast HTML parser from Blink:
https://bugs.chromium.org/p/chromium/issues/detail?id=1407201

This patch adds a new parser that handles a subset of HTML and is only used by
Element::set[Inner|Outer]HTML. While limited, the hope is this parser is
useful for a number of pages. We could expand on the set of cases this handles
in the future.

As a sanity check, there is a debug assertion in place to make sure that the
fast path generates the same output as the full HTML parser would have.

This is a verified 2.25% progression on Speedometer on macOS and 2.60%
progression on iOS.

* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.h:
(WebCore::Document::isDirAttributeDirty const):
(WebCore::Document::setIsDirAttributeDirty):
(WebCore::Document::isTemplateDocument const):
* Source/WebCore/editing/markup.cpp:
(WebCore::createFragmentForMarkup):
* Source/WebCore/html/HTMLBDIElement.cpp:
(WebCore::HTMLBDIElement::HTMLBDIElement):
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::create):
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::childrenChanged):
(WebCore::HTMLElement::dirAttributeChanged):

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::~HTMLInputElement):
(WebCore::HTMLInputElement::needsSuspensionCallback):
We have an optimization in HTMLInputElement which delays the initialization of
m_inputType when constructed by the HTML parser. However, in the fast-parsing
case, we may now fail parsing (due to limited support) in the middle of parsing
an HTMLInputElement. As a result, we may end up destroying an HTMLInputElement
before its m_inputType gets initialized, which would cause a crash. I added a
couple of null checks for m_inputType so that the HTMLInputElement destructor
can run even if m_inputType wasn't initialized yet.

* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::create):
* Source/WebCore/html/HTMLLabelElement.h:
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::create):
* Source/WebCore/html/HTMLSelectElement.h:
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp: Added.
(WebCore::operator==):
(WebCore::onlyContainsLowercaseASCIILetters):
(WebCore::tagNameHash):
(WebCore::HTMLFastPathParser::HTMLFastPathParser):
(WebCore::HTMLFastPathParser::parse):
(WebCore::HTMLFastPathParser::parseResult const):
(WebCore::HTMLFastPathParser::TagInfo::Tag::create):
(WebCore::HTMLFastPathParser::TagInfo::Tag::allowedInPhrasingOrFlowContent):
(WebCore::HTMLFastPathParser::TagInfo::Tag::allowedInFlowContent):
(WebCore::HTMLFastPathParser::TagInfo::ContainerTag::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::ContainsPhrasingContentTag::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::A::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::AWithPhrasingContent::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::B::create):
(WebCore::HTMLFastPathParser::TagInfo::Footer::create):
(WebCore::HTMLFastPathParser::TagInfo::I::create):
(WebCore::HTMLFastPathParser::TagInfo::Input::create):
(WebCore::HTMLFastPathParser::TagInfo::Option::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::Ol::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::Select::parseChild):
(WebCore::HTMLFastPathParser::TagInfo::Strong::create):
(WebCore::HTMLFastPathParser::TagInfo::Ul::parseChild):
(WebCore::HTMLFastPathParser::parseCompleteInput):
(WebCore::HTMLFastPathParser::isWhitespace):
(WebCore::HTMLFastPathParser::isValidUnquotedAttributeValueChar):
(WebCore::HTMLFastPathParser::isValidAttributeNameChar):
(WebCore::HTMLFastPathParser::isCharAfterTagNameOrAttribute):
(WebCore::HTMLFastPathParser::isCharAfterUnquotedAttribute):
(WebCore::HTMLFastPathParser::skipWhitespace):
(WebCore::HTMLFastPathParser::scanText):
(WebCore::HTMLFastPathParser::scanEscapedText):
(WebCore::HTMLFastPathParser::scanTagName):
(WebCore::HTMLFastPathParser::scanAttributeName):
(WebCore::HTMLFastPathParser::scanAttributeValue):
(WebCore::HTMLFastPathParser::scanEscapedAttributeValue):
(WebCore::HTMLFastPathParser::scanHTMLCharacterReference):
(WebCore::HTMLFastPathParser::didFail):
(WebCore::HTMLFastPathParser::peekNext):
(WebCore::HTMLFastPathParser::consumeNext):
(WebCore::HTMLFastPathParser::parseChildren):
(WebCore::HTMLFastPathParser::processAttribute):
(WebCore::HTMLFastPathParser::parseAttributes):
(WebCore::HTMLFastPathParser::parseSpecificElements):
(WebCore::HTMLFastPathParser::parseElement):
(WebCore::HTMLFastPathParser::parseElementAfterTagName):
(WebCore::HTMLFastPathParser::parseContainerElement):
(WebCore::HTMLFastPathParser::parseVoidElement):
(WebCore::canUseFastPath):
(WebCore::tryFastParsingHTMLFragmentImpl):
(WebCore::tryFastParsingHTMLFragment):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.h: Added.
* Source/WebCore/html/parser/HTMLEntityParser.cpp:
(WebCore::appendLegalEntityFor):
* Source/WebCore/html/parser/HTMLEntityParser.h:
* Source/WebCore/html/parser/HTMLNameCache.h:
(WebCore::HTMLNameCache::makeAttributeQualifiedName):
(WebCore::HTMLNameCache::makeQualifiedName):

Canonical link: https://commits.webkit.org/260148@main
